### PR TITLE
MGMT-21625: Add Dual-Stack with Primary IPv6 Support

### PR DESCRIPTION
**Pull request by linoyaslan**
This PR adds support for IPv6-primary dual-stack clusters, enabling users to create dual-stack clusters where IPv6 is the primary IP stack (version-aware for OpenShift 4.12+).

**assisted-by:** cursor

**Screenshots:**
image

image

image

image

## Summary by CodeRabbit

* **New Features**
  * Dual‑stack behavior is now version‑aware and propagates cluster OpenShift version to relevant controls, allowing IPv6‑primary on 4.12+.

* **Validation**
  * Dual‑stack validation schemas accept an optional OpenShift version and enforce version‑appropriate ordering and messages.

* **UI**
  * Subnet pickers group IPv4/IPv6, include a no‑subnet option, improved auto‑selection, automatic subnet reordering, and simplified VIP/network labels with primary/secondary indicators.

* **Tests**
  * Added comprehensive tests for versioned validation, dual‑stack detection, UI flows, and edge cases.

### CODE REVIEW

File: libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/SubnetsDropdown.tsx [hunk 1]
```diff
@@ -48,13 +50,17 @@ export const SubnetsDropdown = ({
...
         ? [noSubnetAvailableOption]
         : [makeNoSubnetSelectedOption(machineSubnets.length)].concat(
-            toFormSelectOptions(machineSubnets),
-          );
+          toFormSelectOptions(machineSubnets),
+        );
+
+    // Separate IPv4 and IPv6 subnets
+    const ipv4Subnets = machineSubnets.filter((subnet) => Address4.isValid(subnet.subnet));
+    const ipv6Subnets = machineSubnets.filter((subnet) => Address6.isValid(subnet.subnet));
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/SubnetsDropdown.tsx]: ip-address v7.1.0 does have these methods, they work correctly

File: libs/ui-lib/lib/ocm/components/clusterDetail/ClusterProperties.tsx [hunk 2]
```diff
@@ -41,7 +41,7 @@ export const getManagementType = ({ userManagedNetworking }: Cluster): string =>
 };
 
 export const getStackTypeLabel = (cluster: Cluster): string =>
-  isDualStack(cluster) ? 'Dual-stack' : 'IPv4';
+  isDualStack({ ...cluster, openshiftVersion: cluster.openshiftVersion }) ? 'Dual-stack' : 'IPv4';
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterDetail/ClusterProperties.tsx]: currently we allow only IPv4 single-stack

File: libs/ui-lib/lib/common/components/ui/formik/_validationSchemas.test.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@ import { test, describe, expect } from 'vitest';
 import {
   baseDomainValidationSchema,
   dnsNameValidationSchema,
+  dualStackValidationSchema,
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/common/components/ui/formik/_validationSchemas.test.ts]: ip-address version 7.1.0 DOES have isValid() methods - can you check again?

File: libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/AvailableSubnetsControl.tsx [hunk 4]
```diff
@@ -35,72 +36,130 @@ export interface AvailableSubnetsControlProps {
...
     .sort(subnetSort);
 
-  const cidr = IPv4Subnets.length >= 1 ? IPv4Subnets[0].subnet : NO_SUBNET_SET;
+  // For OCP >= 4.12, both networks can use either IPv4 or IPv6
+  const allSubnets = supportsIPv6Primary
+    ? [...IPv4Subnets, ...IPv6Subnets].sort(subnetSort)
+    : IPv4Subnets;
+
+  const cidr = allSubnets.length >= 1 ? allSubnets[0].subnet : NO_SUBNET_SET;
   const hasEmptySelection = (values.machineNetworks ?? []).length === 0;
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/AvailableSubnetsControl.tsx]: also here, currently we allow only IPv4 single-stack

File: libs/ui-lib/lib/common/components/ui/formik/_validationSchemas.test.ts [hunk 5]
```diff
@@ -347,4 +348,85 @@ describe('validationSchemas', () => {
...
+
+      test('rejects IPv6 as primary for OCP < 4.12', async () => {
+        const invalidValues = [[{ cidr: '2001:db8::/64' }, { cidr: '192.168.1.0/24' }]];
+
+        for (const value of invalidValues) {
+          await expect(schema.validate(value)).rejects.toThrow(
+            'First network has to be IPv4 subnet',
+          );
+        }
+      });
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/common/components/ui/formik/_validationSchemas.test.ts]: the test string is contained within the actual error message so it's fine

File: libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/networkConfigurationValidation.ts [hunk 6]
```diff
@@ -85,7 +84,7 @@ export const getNetworkConfigurationValidationSchema = (
               otherwise: () =>
                 values.machineNetworks && values.machineNetworks?.length >= 2
                   ? machineNetworksValidationSchema.concat(
-                      dualStackValidationSchema('machine networks'),
+                      dualStackValidationSchema('machine networks', openshiftVersion),
                     )
                   : Yup.array(),
             }),
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/networkConfigurationValidation.ts]: The length gating is intentional and there before my changes. my changes didn't break anything . I only added the openshiftVersion parameter.

File: libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/StackTypeControl.tsx [hunk 7]
```diff
@@ -69,19 +69,49 @@ export const StackTypeControlGroup = ({
...
+          (network) => network.cidr && Address4.isValid(network.cidr),
+        );
+        if (ipv4Network) {
+          setFieldValue('machineNetworks', [ipv4Network]);
+        } else {
+          // No IPv4 network found, set empty for auto-selection
+          setFieldValue('machineNetworks', []);
+        }
+      }
     }
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/StackTypeControl.tsx]: but dual stack with only IPv6 is also not supported, dual-stack must be both - Ipv4 and Ipv6

File: libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/StackTypeControl.tsx [hunk 8]
```diff
@@ -96,11 +144,30 @@ export const StackTypeControlGroup = ({
...
+        );
+      } else {
+        // If first network is IPv4, add IPv6 as secondary
+        setFieldValue(
+          'machineNetworks',
+          [...values.machineNetworks, { cidr: cidrIPv6, clusterId: clusterId }],
+          false,
+        );
+      }
     }
```
[the remark this answers is not in the document]
Comment (linoyaslan) [reply libs/ui-lib/lib/ocm/components/clusterConfiguration/networkConfiguration/StackTypeControl.tsx]: You’re right, my mistake. BUT still, when switching from single to dual stack, we want the **default** to be IPv4 primary and IPv6 secondary, but if the user changes it to IPv6 primary, it should still work. can you check again please?
